src/control/server/ctl_smd_rpc.go

```diff
@@ -185,185 +186,300 @@ func (svc *ControlService) querySmdPools(ctx context.Context, req *ctlpb.SmdQuer
 	return nil
 }
 
-func (svc *ControlService) smdQueryDevice(ctx context.Context, req *ctlpb.SmdQueryReq) (system.Rank, *ctlpb.SmdQueryResp_Device, error) {
+func (svc *ControlService) smdQueryDevice(ctx context.Context, req *ctlpb.SmdQueryReq) (*system.Rank, *ctlpb.SmdQueryResp_SmdDeviceWithHealth, error) {
 	rank := system.NilRank
 	if req.Uuid == "" {
-		return rank, nil, errors.New("empty UUID in device query")
+		return nil, nil, errors.New("empty UUID in device query")
 	}
 
 	resp := new(ctlpb.SmdQueryResp)
 	if err := svc.querySmdDevices(ctx, req, resp); err != nil {
-		return rank, nil, err
+		return nil, nil, err
 	}
 
 	for _, rr := range resp.Ranks {
 		switch len(rr.Devices) {
 		case 0:
 			continue
 		case 1:
+			dev := rr.Devices[0]
+			if dev == nil {
+				return nil, nil, errors.New("nil device in smd query resp")
+			}
+			if dev.Details == nil {
+				return nil, nil, errors.New("device with nil details in smd query resp")
+			}
 			svc.log.Debugf("smdQueryDevice(): uuid %q, rank %d, states %q", req.Uuid,
-				rr.Rank, rr.Devices[0].DevState)
+				rr.Rank, rr.Devices[0].Details.DevState)
 			rank = system.Rank(rr.Rank)
 
-			return rank, rr.Devices[0], nil
+			return &rank, rr.Devices[0], nil
 		default:
-			return rank, nil, errors.Errorf("device query on %s matched multiple devices", req.Uuid)
+			return nil, nil, errors.Errorf("device query on %s matched multiple devices", req.Uuid)
 		}
 	}
 
-	return rank, nil, nil
+	return &rank, nil, nil
 }
 
-func (svc *ControlService) smdSetFaulty(ctx context.Context, req *ctlpb.SmdQueryReq) (*ctlpb.SmdQueryResp, error) {
+func (svc *ControlService) smdGetEngine(ctx context.Context, req *ctlpb.SmdQueryReq) (*system.Rank, *Engine, error) {
 	req.Rank = uint32(system.NilRank)
 	rank, device, err := svc.smdQueryDevice(ctx, req)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	if device == nil {
-		return nil, errors.Errorf("smdSetFaulty on %s did not match any devices", req.Uuid)
+		return nil, nil, errors.Errorf("%s did not match any devices", req.Uuid)
 	}
 
```

Comment:
We could do but even if the check is not done here an empty list of engine instances will be returned from `FilterInstancesByRankSet()` and the function will return error on L242 printing the bogus rank #. I agree that it would be an improvement to fail earlier so I will update if I have to repush.

Reply:
fair enough